multicluster/cmd/multicluster-controller/memberclusterannounce_webhook.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2021 Antrea Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"context"
+	"fmt"
+	serviceaccount "k8s.io/apiserver/pkg/authentication/serviceaccount"
+	"k8s.io/klog/v2"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// log is for logging in this package.
+var memberclusterannouncelog = logf.Log.WithName("memberclusterannounce-resource")
+
+//+kubebuilder:webhook:path=/validate-multicluster-crd-antrea-io-v1alpha1-memberclusterannounce,mutating=false,failurePolicy=fail,sideEffects=None,groups=multicluster.crd.antrea.io,resources=memberclusterannounces,verbs=create;update,versions=v1alpha1,name=vmemberclusterannounce.kb.io,admissionReviewVersions={v1,v1beta1}
+
+// member cluster announce validator
+type memberClusterAnnounceValidator struct {
+	Client    client.Client
+	decoder   *admission.Decoder
+	namespace string
+}
+
+// Handle handles admission requests.
+func (v *memberClusterAnnounceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	memberClusterAnnounce := &multiclusterv1alpha1.MemberClusterAnnounce{}
+	klog.InfoS("Received admission request", "req", req)
+	e := v.decoder.Decode(req, memberClusterAnnounce)
+	if e != nil {
+		klog.ErrorS(e, "error while decoding")
+		return admission.Errored(http.StatusBadRequest, e)
+	}
+	klog.InfoS("Received", "memberClusterAnnounce", memberClusterAnnounce)
+
+	ui := req.UserInfo
+	_, saName, err := serviceaccount.SplitUsername(ui.Username)
+	if err != nil {
+		klog.ErrorS(err, "Error getting service account name", "request", req)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// read the ClusterSet info
+	clusterSetList := &multiclusterv1alpha1.ClusterSetList{}
+	if err := v.Client.List(context.TODO(), clusterSetList, client.InNamespace(v.namespace)); err != nil {
+		klog.ErrorS(err, "Error reading ClusterSet", "Namespace", v.namespace)
+		return admission.Errored(http.StatusPreconditionFailed, err)
+	}
+
+	if len(clusterSetList.Items) != 1 {
+		klog.InfoS("Invalid ClusterSet config", "Namespace", v.namespace)
+		return admission.Errored(http.StatusPreconditionFailed, fmt.Errorf("Invalid clusterSet config"))
```

Comment:
s/clusterSet/ClusterSet/

Reply:
Can `Invalid clusterSet config` be more specific? I think it's a little misleading since user might think it's something wrong in the request.

Reply:
maybe `Invalid ClusterSet config in the elected leader cluster, please contact your administrator`?

Reply:
Done, thanks for pointing it out.